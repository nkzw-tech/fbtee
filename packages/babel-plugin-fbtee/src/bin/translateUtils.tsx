import { readFileSync } from 'node:fs';
import { pathToFileURL } from 'node:url';
import FbtHashKey from '../fbtHashKey.tsx';
import nullthrows from '../nullthrows.tsx';
import { FbtSite } from '../translate/FbtSite.tsx';
import type {
  HashToTranslation,
  TranslationResult,
} from '../translate/TranslationBuilder.tsx';
import TranslationBuilder from '../translate/TranslationBuilder.tsx';
import TranslationConfig from '../translate/TranslationConfig.tsx';
import type { SerializedTranslationData } from '../translate/TranslationData.tsx';
import TranslationData from '../translate/TranslationData.tsx';
import type { PatternHash, PatternString } from '../Types.ts';
import type { CollectFbtOutput, CollectFbtOutputPhrase } from './collect.tsx';

export type Options = Readonly<{
  hashModule: boolean | string;
  jenkins: boolean;
  strict: boolean;
}>;

export type LocaleToHashToTranslationResult = {
  [fbLocale: string]: Partial<Record<PatternHash, TranslationResult>>;
};

/** Phrases translated for a specific locale */
type TranslatedGroup = Readonly<{
  ['fb-locale']: string;
  translatedPhrases: ReadonlyArray<TranslationResult>;
}>;

export type TranslatedGroups = ReadonlyArray<TranslatedGroup>;

/** Translations in a specific locale */
export type TranslationGroup = Readonly<{
  ['fb-locale']: string;
  translations: Translations;
}>;

export type Translations = Partial<
  Record<PatternString, SerializedTranslationData | null>
>;

/** Phrases and translation data in one JSON object */
type InputJSONType = Readonly<{
  phrases: ReadonlyArray<CollectFbtOutputPhrase>;
  translationGroups: ReadonlyArray<TranslationGroup>;
}>;

export function loadJSON<T>(filepath: string): T {
  try {
    return JSON.parse(readFileSync(filepath).toString());
  } catch (error) {
    if (error instanceof Error) {
      error.message += `\nFile path: "${filepath}"`;
    }
    throw error;
  }
}

export async function processFiles(
  stringFile: string,
  translationFiles: ReadonlyArray<string>,
  options: Options,
): Promise<LocaleToHashToTranslationResult> {
  const { phrases } = loadJSON<CollectFbtOutput>(stringFile);
  const fbtSites = phrases.map(createFbtSiteFromJSON);
  return await processGroups(
    phrases,
    translationFiles.map((file) =>
      processTranslations(fbtSites, loadJSON<TranslationGroup>(file), options),
    ),
    options,
  );
}

export async function processJSON(
  json: InputJSONType,
  options: Options,
): Promise<LocaleToHashToTranslationResult | TranslatedGroups> {
  const fbtSites = json.phrases.map(createFbtSiteFromJSON);
  return await processGroups(
    json.phrases,
    json.translationGroups.map((group) =>
      processTranslations(fbtSites, group, options),
    ),
    options,
  );
}

async function processGroups(
  phrases: ReadonlyArray<CollectFbtOutputPhrase>,
  translatedGroups: TranslatedGroups,
  options: Options,
): Promise<LocaleToHashToTranslationResult> {
  let fbtHash: typeof FbtHashKey | null = null;
  if (options.jenkins) {
    fbtHash = (await import('../fbtHashKey.tsx')).default;
  } else if (typeof options.hashModule === 'string') {
    fbtHash = (await import(pathToFileURL(options.hashModule).href)).default;
  }

  if (!fbtHash) {
    return translatedGroups as unknown as LocaleToHashToTranslationResult;
  }

  const localeToHashToFbt: LocaleToHashToTranslationResult = {};
  for (const group of translatedGroups) {
    const hashToFbt: Partial<Record<string, TranslationResult>> =
      (localeToHashToFbt[group['fb-locale']] = {});
    phrases.forEach((phrase, idx) => {
      const translatedFbt = group.translatedPhrases[idx];
      const jsfbt = nullthrows(
        phrase.jsfbt,
        `Expect every phrase to have 'jsfbt' field. However, 'jsfbt' is missing in the phrase at index ${idx}.`,
      );
      const hash = nullthrows(fbtHash)(jsfbt.t);
      hashToFbt[hash] = translatedFbt;
    });
  }
  return localeToHashToFbt;
}

function checkAndFilterTranslations(
  locale: string,
  translations: Translations,
  options: Options,
): Translations {
  const filteredTranslations: Translations = {};
  for (const hash of Object.keys(translations)) {
    if (translations[hash] == null) {
      const message = `Missing ${locale} translation for string (${hash})`;
      if (options.strict) {
        throw new Error(message);
      } else {
        console.error(message);
      }
    } else {
      filteredTranslations[hash] = translations[hash];
    }
  }
  return filteredTranslations;
}

function processTranslations(
  fbtSites: ReadonlyArray<FbtSite>,
  group: TranslationGroup,
  options: Options,
): TranslatedGroup {
  const config = TranslationConfig.fromFBLocale(group['fb-locale']);
  const filteredTranslations = checkAndFilterTranslations(
    group['fb-locale'],
    group.translations,
    options,
  );
  const translations: HashToTranslation = {};
  for (const t of Object.keys(filteredTranslations)) {
    translations[t] = TranslationData.fromJSON(filteredTranslations[t]);
  }
  const translatedPhrases = fbtSites.map((fbtsite) =>
    new TranslationBuilder(translations, config, fbtsite, false).build(),
  );
  return {
    'fb-locale': group['fb-locale'],
    translatedPhrases,
  };
}

function createFbtSiteFromJSON(json: CollectFbtOutputPhrase): FbtSite {
  return FbtSite.fromScan(json);
}
