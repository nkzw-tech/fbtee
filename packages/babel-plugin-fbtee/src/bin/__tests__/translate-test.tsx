import { afterEach, describe, it, jest } from '@jest/globals';
import { jsCodeNonASCIICharSerializer } from '../../__tests__/FbtTestUtil.tsx';
import { Options, processJSON } from '../translateUtils.tsx';

expect.addSnapshotSerializer(jsCodeNonASCIICharSerializer);

const consoleError = console.error;

afterEach(() => {
  console.error = consoleError;
});

function testTranslateNewPhrases(options: Options) {
  it('should not throw on missing translations', async () => {
    console.error = jest.fn();
    const result = await processJSON(
      {
        phrases: [
          {
            col_beg: 12,
            col_end: 49,
            filepath: 'src/example/Example.react.js',
            hashToLeaf: {
              '2dcba29d4a842c6be5d76fe996fcd9f4': {
                desc: 'title',
                text: 'Your FBT Demo',
              },
            },
            jsfbt: {
              m: [],
              t: {
                desc: 'title',
                text: 'Your FBT Demo',
                tokenAliases: {},
              },
            },
            line_beg: 130,
            line_end: 130,
            project: 'fbt-demo-project',
          },
        ],
        translationGroups: [
          {
            'fb-locale': 'fb_HX',
            translations: {
              '2dcba29d4a842c6be5d76fe996fcd9f4': null,
            },
          },
        ],
      },
      options,
    );
    expect(result).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });

  it('should translate string with no variation', async () => {
    const result = await processJSON(
      {
        phrases: [
          {
            col_beg: 12,
            col_end: 49,
            filepath: 'src/example/Example.react.js',
            hashToLeaf: {
              '2dcba29d4a842c6be5d76fe996fcd9f4': {
                desc: 'title',
                text: 'Your FBT Demo',
              },
            },
            jsfbt: {
              m: [],
              t: {
                desc: 'title',
                text: 'Your FBT Demo',
                tokenAliases: {},
              },
            },
            line_beg: 130,
            line_end: 130,
            project: 'fbt-demo-project',
          },
        ],
        translationGroups: [
          {
            'fb-locale': 'fb_HX',
            translations: {
              '2dcba29d4a842c6be5d76fe996fcd9f4': {
                tokens: [],
                translations: [
                  {
                    translation: 'Translation data for Your FBT Demo',
                    variations: {},
                  },
                ],
                types: [],
              },
            },
          },
        ],
      },
      options,
    );
    expect(result).toMatchSnapshot();
  });

  it('should translate string with variations and inner strings', async () => {
    const result = await processJSON(
      {
        phrases: [
          {
            col_beg: 8,
            col_end: 14,
            filepath: 'src/example/Example.react.js',
            hashToLeaf: {
              'PqPPir8Kg9xSlqdednPFOg==': {
                desc: 'example 1',
                text: '{name} has shared {=a photo} with you. View a photo',
              },
              'gVKMc/8jq5vnYR5v2bb32g==': {
                desc: 'example 1',
                text: '{name} has shared {=[number] photos} with you. View =[number] photos',
              },
            },
            jsfbt: {
              m: [
                {
                  token: 'name',
                  type: 1,
                },
                {
                  singular: true,
                  token: 'number',
                  type: 2,
                },
              ],
              t: {
                '*': {
                  '*': {
                    desc: 'example 1',
                    text: '{name} has shared {=[number] photos} with you. View =[number] photos',
                    tokenAliases: {
                      '=[number] photos': '=m2',
                    },
                  },
                  _1: {
                    desc: 'example 1',
                    text: '{name} has shared {=a photo} with you. View a photo',
                    tokenAliases: {
                      '=a photo': '=m2',
                    },
                  },
                },
              },
            },
            line_beg: 127,
            line_end: 142,
            project: 'fbt-demo-project',
          },
          {
            col_beg: 10,
            col_end: 14,
            filepath: 'src/example/Example.react.js',
            hashToLeaf: {
              '/gj3gwqx1z8Xw233oZgOpQ==': {
                desc: 'In the phrase: "{name} has shared {=[number] photos} with you. View =[number] photos"',
                text: '{number} photos',
              },
              '8UZCD6gFUKN+U5UUo1I3/w==': {
                desc: 'In the phrase: "{name} has shared {=a photo} with you. View a photo"',
                text: 'a photo',
              },
            },
            jsfbt: {
              m: [
                {
                  token: 'name',
                  type: 1,
                },
                {
                  singular: true,
                  token: 'number',
                  type: 2,
                },
              ],
              t: {
                '*': {
                  '*': {
                    desc: 'In the phrase: "{name} has shared {=[number] photos} with you. View =[number] photos"',
                    text: '{number} photos',
                    tokenAliases: {},
                  },
                  _1: {
                    desc: 'In the phrase: "{name} has shared {=a photo} with you. View a photo"',
                    text: 'a photo',
                    tokenAliases: {},
                  },
                },
              },
            },
            line_beg: 133,
            line_end: 140,
            project: 'fbt-demo-project',
          },
        ],
        translationGroups: [
          {
            'fb-locale': 'fb_HX',
            translations: {
              '/gj3gwqx1z8Xw233oZgOpQ==': {
                tokens: [],
                translations: [
                  {
                    id: 107_911_344,
                    translation: 'translation is: {number} photos',
                    variations: {},
                  },
                ],
                types: [],
              },
              '8UZCD6gFUKN+U5UUo1I3/w==': {
                tokens: [],
                translations: [
                  {
                    id: 107_911_340,
                    translation: 'translation is: a photo',
                    variations: {},
                  },
                ],
                types: [],
              },
              'PqPPir8Kg9xSlqdednPFOg==': {
                tokens: ['name'],
                translations: [
                  {
                    id: 108_537_963,
                    translation:
                      'translation is: {name} has shared {=a photo}. View a photo',
                    variations: { '0': 2 },
                  },
                  {
                    id: 108_537_953,
                    translation:
                      'translation is: {name} has shared {=a photo}. View a photo',
                    variations: { '0': 1 },
                  },
                  {
                    id: 108_537_972,
                    translation:
                      'translation is: {name} has shared {=a photo}. View a photo',
                    variations: { '0': 3 },
                  },
                ],
                types: [3],
              },
              'gVKMc/8jq5vnYR5v2bb32g==': {
                tokens: ['name'],
                translations: [
                  {
                    id: 108_537_963,
                    translation:
                      'translation is: {name} has shared {=[number] photos}. View =[number] photos',
                    variations: { '0': 2 },
                  },
                  {
                    id: 108_537_953,
                    translation:
                      'translation is: {name} has shared {=[number] photos}. View =[number] photos',
                    variations: { '0': 1 },
                  },
                  {
                    id: 108_537_972,
                    translation:
                      'translation is: {name} has shared {=[number] photos}. View =[number] photos',
                    variations: { '0': 3 },
                  },
                ],
                types: [3],
              },
            },
          },
        ],
      },
      options,
    );
    expect(result).toMatchSnapshot();
  });

  it('should translate string with metadata entries that create no hidden variation. Note: this string was collected in RN mode.', async () => {
    const result = await processJSON(
      {
        phrases: [
          {
            col_beg: 6,
            col_end: 12,
            filepath: 'src/example/Example.react.js',
            hashToLeaf: {
              'j9fTl1uOEIuslim41sMkdQ==': {
                desc: 'Example enum',
                text: 'he shared a photo.',
              },
              'sNncqVnQfCGCeJNXsLObVw==': {
                desc: 'Example enum',
                text: 'they shared a photo.',
              },
              'vHtEb4ph7GJGeRkjtEHcPA==': {
                desc: 'Example enum',
                text: 'she shared a photo.',
              },
            },
            jsfbt: {
              m: [
                {
                  type: 3,
                },
              ],
              t: {
                '*': {
                  desc: 'Example enum',
                  text: 'they shared a photo.',
                },
                '1': {
                  desc: 'Example enum',
                  text: 'she shared a photo.',
                },
                '2': {
                  desc: 'Example enum',
                  text: 'he shared a photo.',
                },
              },
            },
            line_beg: 127,
            line_end: 130,
            project: 'fbt-demo-project',
          },
        ],
        translationGroups: [
          {
            'fb-locale': 'fb_HX',
            translations: {
              'j9fTl1uOEIuslim41sMkdQ==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: he shared a photo',
                    variations: {},
                  },
                ],
                types: [],
              },
              'sNncqVnQfCGCeJNXsLObVw==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: they shared a photo',
                    variations: {},
                  },
                ],
                types: [],
              },
              'vHtEb4ph7GJGeRkjtEHcPA==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: she shared a photo',
                    variations: {},
                  },
                ],
                types: [],
              },
            },
          },
        ],
      },
      options,
    );
    expect(result).toMatchSnapshot();
  });

  it('should translate string with enum', async () => {
    const result = await processJSON(
      {
        phrases: [
          {
            col_beg: 6,
            col_end: 12,
            filepath: 'src/example/Example.react.js',
            hashToLeaf: {
              '/3R5GnCZ5eU3EgRAiLf1vA==': {
                desc: 'Example enum',
                text: '{name} has a photo to share! View photo.',
              },
              '/giEGYE5cqdJVvtszgdPLg==': {
                desc: 'Example enum',
                text: '{name} has a video to share! View video.',
              },
              '2PhpGvvUtmT5RTpv8Kqf0w==': {
                desc: 'Example enum',
                text: '{name} has a link to share! View link.',
              },
              'nwcWZzo5dAQX38+P1IaY6A==': {
                desc: 'Example enum',
                text: '{name} has a page to share! View page.',
              },
              'wGYWno21D5FWihP/v0boFw==': {
                desc: 'Example enum',
                text: '{name} has a post to share! View post.',
              },
            },
            jsfbt: {
              m: [null],
              t: {
                LINK: {
                  desc: 'Example enum',
                  text: '{name} has a link to share! View link.',
                },
                PAGE: {
                  desc: 'Example enum',
                  text: '{name} has a page to share! View page.',
                },
                PHOTO: {
                  desc: 'Example enum',
                  text: '{name} has a photo to share! View photo.',
                },
                POST: {
                  desc: 'Example enum',
                  text: '{name} has a post to share! View post.',
                },
                VIDEO: {
                  desc: 'Example enum',
                  text: '{name} has a video to share! View video.',
                },
              },
            },
            line_beg: 127,
            line_end: 133,
            project: 'fbt-demo-project',
          },
        ],
        translationGroups: [
          {
            'fb-locale': 'fb_HX',
            translations: {
              '/3R5GnCZ5eU3EgRAiLf1vA==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: {name} has a photo to share',
                    variations: {},
                  },
                ],
                types: [],
              },
              '/giEGYE5cqdJVvtszgdPLg==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: {name} has a video to share',
                    variations: {},
                  },
                ],
                types: [],
              },
              '2PhpGvvUtmT5RTpv8Kqf0w==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: {name} has a link to share',
                    variations: {},
                  },
                ],
                types: [],
              },
              'nwcWZzo5dAQX38+P1IaY6A==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: {name} has a page to share',
                    variations: {},
                  },
                ],
                types: [],
              },
              'wGYWno21D5FWihP/v0boFw==': {
                tokens: [],
                translations: [
                  {
                    translation: 'translation is: {name} has a post to share',
                    variations: {},
                  },
                ],
                types: [],
              },
            },
          },
        ],
      },
      options,
    );
    expect(result).toMatchSnapshot();
  });
}

describe('translate-test.js', () => {
  describe('should translate new jsfbt payload', () => {
    for (const options of [
      { hashModule: false, jenkins: false, strict: false } as const,
    ]) {
      describe(`with option=${JSON.stringify(options)}:`, () => {
        testTranslateNewPhrases(options);
      });
    }
  });
});
