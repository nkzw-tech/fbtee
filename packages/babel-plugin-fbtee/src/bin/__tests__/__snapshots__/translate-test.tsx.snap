// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`translate-test.js should translate new jsfbt payload with option={"hashModule":false,"jenkins":false,"strict":false}: should not throw on missing translations 1`] = `
[
  {
    "fb-locale": "fb_HX",
    "translatedPhrases": [
      "Your FBT Demo",
    ],
  },
]
`;

exports[`translate-test.js should translate new jsfbt payload with option={"hashModule":false,"jenkins":false,"strict":false}: should translate string with enum 1`] = `
[
  {
    "fb-locale": "fb_HX",
    "translatedPhrases": [
      {
        "LINK": "translation is: {name} has a link to share",
        "PAGE": "translation is: {name} has a page to share",
        "PHOTO": "translation is: {name} has a photo to share",
        "POST": "translation is: {name} has a post to share",
        "VIDEO": "translation is: {name} has a video to share",
      },
    ],
  },
]
`;

exports[`translate-test.js should translate new jsfbt payload with option={"hashModule":false,"jenkins":false,"strict":false}: should translate string with metadata entries that create no hidden variation. Note: this string was collected in RN mode. 1`] = `
[
  {
    "fb-locale": "fb_HX",
    "translatedPhrases": [
      {
        "*": "translation is: they shared a photo",
        "1": "translation is: she shared a photo",
        "2": "translation is: he shared a photo",
      },
    ],
  },
]
`;

exports[`translate-test.js should translate new jsfbt payload with option={"hashModule":false,"jenkins":false,"strict":false}: should translate string with no variation 1`] = `
[
  {
    "fb-locale": "fb_HX",
    "translatedPhrases": [
      "Translation data for Your FBT Demo",
    ],
  },
]
`;

exports[`translate-test.js should translate new jsfbt payload with option={"hashModule":false,"jenkins":false,"strict":false}: should translate string with variations and inner strings 1`] = `
[
  {
    "fb-locale": "fb_HX",
    "translatedPhrases": [
      {
        "*": {
          "*": "translation is: {name} has shared {=m2}. View =[number] photos",
          "_1": "translation is: {name} has shared {=m2}. View a photo",
        },
      },
      {
        "*": {
          "*": "translation is: {number} photos",
          "_1": "translation is: a photo",
        },
      },
    ],
  },
]
`;
